### PR TITLE
original post discovery v2.0

### LINKED ISSUE #51: web site button layout broken in one width range
e.g. 1024px.

screen shot 2015-11-04 at 10 49 48 am

### PR DESCRIPTION
**Pull request by snarfed**
...specifically, implements much of https://github.com/snarfed/bridgy/issues/51#issuecomment-136018857 (but not all). corresponds to snarfed/bridgy#465.

### CODE REVIEW

File: granary/source.py [hunk 1]
```diff
@@ -444,28 +457,35 @@ def article_urls(field):
...
-    tags = article_urls('tags')
-    urls = attachments | set(util.extract_links(content))
+
+    originals = set()
+    links = util.extract_links(content)
+    if links:
+      last = links[-1]
+      if (content.find(last) + len(last) >= len(content) - 4 and
+          (not domains or util.domain_from_link(last) in domains)):
+        originals.add(links.pop())
```
Comment (kylewm) [inline granary/source.py]: cool

File: granary/source.py [hunk 2]
```diff
@@ -444,28 +457,35 @@ def article_urls(field):
...
-      if http not in uds and https not in uds and not ellipsized(http):
-        uds.append(http)
+      if http not in originals and https not in originals and not ellipsized(http):
+        originals.add(http)
 
     obj.setdefault('tags', []).extend(
       {'objectType': 'article', 'url': u} for u in urls if not ellipsized(u))
+    if originals:
+      uds = obj.setdefault('upstreamDuplicates', [])
+      uds.extend(o for o in originals if o not in uds)
```
Comment (kylewm) [inline granary/source.py]: makes sense

File: granary/test/test_source.py [hunk 3]
```diff
@@ -82,76 +82,80 @@ def setUp(self):
...
+      'tags': [{'objectType': 'article', 'url': 'http://bar/2'}],
+    })
+    check(obj, tags=['http://foo/1', 'http://bar/2', 'http://baz/3'])
+
+    # links at the end (modulo punctuation) become upstreamDuplicates
+    for end in '', ' ', '.' ')', ') ', ' ! ', ' :-D':
+      check({'content': 'asdf http://middle http://end' + end},
+            uds=['http://end'], tags=['http://middle'])
+
+    check({'content': 'asdf http://too far!'}, tags=['http://too'])
```
Comment (kylewm) [inline granary/test/test_source.py]: heh nice

File: granary/facebook.py [hunk 4]
```diff
@@ -852,14 +852,11 @@ def post_to_object(self, post, _type='post'):
...
-    obj['tags'] += [self.postprocess_object({
-      'objectType': 'article',
-      'url': act.get('link'),
-      'displayName': act.get('name')
-    }) for act in see_orig_actions]
+    see_orig_links = filter(
+      None, (act.get('link') for act in post.get('actions',[])
+             if act.get('name', '').lower() in SEE_ORIGINAL_ACTIONS))
+    if see_orig_links:
+      obj.setdefault('upstreamDuplicates', []).extend(see_orig_links)
```
Comment (kylewm) [inline granary/facebook.py]: I'm very confused how "See Original"s weren't already counted as upstreamDuplicates. Did they get upgraded later or something?

Comment (snarfed) [inline granary/facebook.py]: ah. iiuc we used to only use `upstreamDuplicates` for `syndication` links. with these PRs, we use them for all links that we decide are POSSE original posts.

Comment (kylewm) [inline granary/facebook.py]: ahhh, yeah that makes sense. cool, I like that a lot.

**Pull request merged**